new whip server

j0sh opened the pull request:
Introduces a WHIP endpoint at `/live/video-to-video/{streamid}/whip`

This strictly affects ingest only; it does not touch selection, trickle or anything beyond those.

There should also be no changes to the existing MediaMTX whip / rtmp ingest. The old ingest is safe to run alongside this and there should be no behavioral changes to that. The one modification is to tighten up request methods in the HTTP handler (adding `POST` prefixes etc) to prevent route conflicts with the new WHIP endpoints.

### New Configuration Options

Currently implemented directly as environment variables.

- LIVE_AI_WHIP_ADDR: UDP binding address for WebRTC (default: 0.0.0.0:7290)                                                                                                                                                                                  
- LIVE_AI_NAT_IP: Public IP for NAT traversal (optional)                                                                                                                                                                                                     
- LIVE_AI_VIDEO_GATHER_TIMEOUT: Timeout for gathering video tracks (default: 5 seconds)                                                                                                                                                                      
- LIVE_AI_PLAYBACK_HOST: RTMP URL for output (default: rtmp://localhost/)                                                                                                                                                                                    
 - LIVE_AI_ALLOW_CORS: Enable CORS headers for browser clients (optional)                                                                                                                                                                                     

### Other Changes

Added a `clog.Info` function for a slog-style logging API

### Not Yet Implemented
- Proper handling + documentation for the new configuration vars
- ICE restarts + trickle ICE
- DELETE
- TCP, TURN, etc

These todos are all in Linear for later follow-up.

j0sh committed:
Initial WHIP server

j0sh committed:
lint

j0sh committed:
Make whip conditional to fix tests

j0sh commented:
Added the ability to disable WHIP ingest entirely by omitting the `LIVE_AI_WHIP_ADDR` env var. This was mostly done to fix tests but may be useful if we need to disable whip for whatever reason, without a roll back. https://github.com/livepeer/go-livepeer/pull/3432/commits/522d9e69c14f7bf0dad93685ac840896cf5a4549

Ideally we'd have a CLI argument like  `-whipIngest` but that can come later.

j0sh committed:
remove more leftovers

leszko commented on clog/clog.go:
```diff
@@ -193,6 +193,48 @@ func infof(ctx context.Context, lastErr bool, publicLog bool, format string, arg
+// Info logs a message with key-value pairs in a slog-like style.
+// Example: Info(ctx, "hello", "key1", value1, "key2", value2)
+// This will log: "hello key1=value1 key2=value2"
+func Info(ctx context.Context, msg string, keyvals ...interface{}) {
```
I wonder if we don't make our `clog.go` more confusing with this change. Normally, you'd think that `Info()` is the same as `Infof()` but without the formatting values inside. However, this implementation makes `Info()` behaves totally different than `Infof()`. 

@mjh1 Wdyt? You recently did some work re:logging.

j0sh replied:
Not sure that I understand the concern. The behavior matches that of `slog.Info` . The `f` convention to indicate a format string is used elsewhere in golang, eg log.Info / Infof, fmt.Errorf, etc. We are not doing anything different here. Personally, I like the brevity of plain `Info` and consistency / ease of use compared to munging format strings.

mjh1 replied:
Yeah I kind of see the concern it's two different syntaxes when usually libraries seem to support one or the other, but for our case I think it's ok. We want to be able to use the newer style syntax, but also want to be able to use our existing context implementation. I think the only alternative really is to implement our contextual stuff in slog which is what I did [here](https://github.com/livepeer/go-livepeer/pull/3412/commits/2392bd41a214180b7522a3a9f9e9473b0f7f8f81), but that ends up being a bit  confusing as some places we're using slog rather than clog.
The other thing is that this method can still be used as you described, just for a simple log line without values, just don't pass any keyvals.

leszko commented on server/ai_mediaserver.go:
```diff
@@ -88,11 +89,20 @@ func startAIMediaServer(ctx context.Context, ls *LivepeerServer) error {
 
+	// Configure WHIP ingest only if an addr is specified.
+	// TODO use a proper cli flag
+	if os.Getenv("LIVE_AI_WHIP_ADDR") != "" {
```
How about moving this to flags? I think we don't use anywhere the env variables directly in go-livepeer, but just adding more flags.

j0sh replied:
Yes it has been mentioned in several places this can be done as a follow up to this PR

leszko commented on server/ai_mediaserver.go:
```diff
@@ -687,6 +697,154 @@ func (ls *LivepeerServer) GetLiveVideoToVideoStatus() http.Handler {
+					pipeline = authResp.Pipeline
+				}
+
+				if len(authResp.paramsMap) > 0 {
```
I see there is a lot of code copied from `StartLiveVideo()`, I'd consider refactoring it and re-use. It may be okeyish to just copy-paste the code like you did, but I think there is some functionality missing (e.g. I don't see sending `stream_trace`) into Kafka, which may make it more difficult to analyze.

j0sh replied:
Can add the stream_trace (I think that was added after we wrote this code) but wanted to avoid modifying existing ingest code paths - they are too different and the risk was too high of breaking something

leszko replied:
ok, then let's just add the `stream_trace`.

j0sh replied:
Added stream_trace and some other missing metrics in https://github.com/livepeer/go-livepeer/pull/3432/commits/7236f832e9ca979a8c2aa9e76cc28aa2c35ad6ca

leszko commented on media/whip_server.go:
```diff
@@ -0,0 +1,650 @@
+			PayloadType:        105,
+		},
+
+		// TODO verify 42e01f won't produle bframes
```
Let's remove this before merging

j0sh replied:
Sure, this was part of the default set of Pion codecs but I don't think this part even does what we need it to right now - the remote SDPs the browser is setting don't look as constrained as I was expecting. Needs more investigation

j0sh replied:
Removed in https://github.com/livepeer/go-livepeer/pull/3432/commits/a5633d6f17980cd995e52f155e2b1e25c7f74af8

leszko reviewed (commented):
Great work @j0sh Added some comments, mostly to the ai_mediaserver.go part. I was not able to review in depth the media part, in particular the `whip_server.go`.

My suggestion is to address the comments and get it merged, then we can test it in action:
1. I'd not use env vars, but move it to flags (as we do with everything else)
2. It'd be super beneficial to describe how you test this whip endpoint locally (can be a README or even some short Notion doc)

j0sh committed:
Add missing metrics

j0sh committed:
Avoid blocking WHIP response if orch selection is blocked

j0sh committed:
Packet queue limit in RTPSegmenter

j0sh committed:
Reject non-H.264 video codecs

j0sh committed:
Remove unused codec registrations

j0sh commented:
Addressed PR feedback, and added the following:

Makes WHIP not be blocked by slow (or broken) orch selection: https://github.com/livepeer/go-livepeer/pull/3432/commits/147af6b3b1287b2a9d3b0281c2e5d3da6f6bd41f 

Adds a packet queue limit in the segmenter: https://github.com/livepeer/go-livepeer/pull/3432/commits/1203f5a88b9e642136f549a5c0145578f878710e

Reject non-H.264 video codecs: https://github.com/livepeer/go-livepeer/pull/3432/commits/3c10a601ab30e03daf48b628630df9c90991eff8

j0sh committed:
Include ICE servers in Link header for OPTIONS

j0sh committed:
Merge branch 'master' into ja/whip-server

j0sh committed:
Remove unnecessary http response from auth failure

mjh1 committed:
Merge branch 'master' into ja/whip-server

j0sh committed:
Merge branch 'master' into ja/whip-server
